// battle-node-v2/dist/protocol/PacketBuilder.js
import { crc32 } from '../utils/crc32.js';
import { PACKET_HEADER, FF_MARKER, PacketType } from './Constants.js';
export class PacketBuilder {
    static buildLogin(password) {
        const body = Buffer.from(password, 'utf8');
        return PacketBuilder.constructPacket(PacketType.Login, body);
    }
    static buildCommand(sequence, command) {
        const cmdBuffer = Buffer.from(command, 'utf8');
        const body = Buffer.alloc(1 + cmdBuffer.length);
        body.writeUInt8(sequence, 0);
        cmdBuffer.copy(body, 1);
        return PacketBuilder.constructPacket(PacketType.Command, body);
    }
    static buildKeepAlive(sequence) {
        const body = Buffer.alloc(1);
        body.writeUInt8(sequence, 0);
        return PacketBuilder.constructPacket(PacketType.Command, body);
    }
    static buildAcknowledge(sequence) {
        const body = Buffer.alloc(1);
        body.writeUInt8(sequence, 0);
        return PacketBuilder.constructPacket(PacketType.Message, body);
    }
    static constructPacket(type, body) {
        const payloadHeader = Buffer.from([FF_MARKER, type]);
        const crcInput = Buffer.concat([payloadHeader, body]);
        const checksum = crc32(crcInput);
        const packet = Buffer.alloc(6 + crcInput.length);
        PACKET_HEADER.copy(packet, 0);
        packet.writeInt32LE(checksum, 2);
        crcInput.copy(packet, 6);
        return packet;
    }
}
