// battle-node-v2/dist/transport/TcpTransport.js
import * as net from 'net';
import { EventEmitter } from 'events';
export class TcpTransport extends EventEmitter {
    socket;
    config;
    isConnected = false;
    constructor(config) {
        super();
        this.config = config;
        this.socket = new net.Socket();
        this.setupListeners();
    }
    setupListeners() {
        this.socket.on('data', (data) => {
            this.emit('message', data);
        });
        this.socket.on('error', (err) => {
            this.emit('error', err);
        });
        this.socket.on('close', () => {
            this.isConnected = false;
            this.emit('close');
        });
        this.socket.on('connect', () => {
            this.isConnected = true;
            this.emit('connected');
        });
    }
    async connect() {
        return new Promise((resolve, reject) => {
            const errorHandler = (err) => {
                this.socket.removeListener('connect', connectHandler);
                reject(err);
            };
            const connectHandler = () => {
                this.socket.removeListener('error', errorHandler);
                resolve();
            };
            this.socket.once('error', errorHandler);
            this.socket.once('connect', connectHandler);
            this.socket.connect(this.config.port, this.config.ip);
        });
    }
    async send(data) {
        return new Promise((resolve, reject) => {
            if (!this.isConnected) {
                return reject(new Error('Socket not connected'));
            }
            const success = this.socket.write(data, (err) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
            if (!success) {
            }
        });
    }
    close() {
        if (!this.socket.destroyed) {
            this.socket.destroy();
        }
    }
}
