// battle-node-v2/dist/transport/UdpTransport.js
import * as dgram from 'dgram';
import { EventEmitter } from 'events';
export class UdpTransport extends EventEmitter {
    socket;
    config;
    isConnected = false;
    constructor(config) {
        super();
        this.config = config;
        this.socket = dgram.createSocket('udp4');
        this.setupListeners();
    }
    setupListeners() {
        this.socket.on('message', (msg, rinfo) => {
            if (rinfo.address === this.config.ip && rinfo.port === this.config.port) {
                this.emit('message', msg);
            }
        });
        this.socket.on('error', (err) => {
            this.emit('error', err);
        });
        this.socket.on('close', () => {
            this.isConnected = false;
            this.emit('close');
        });
    }
    async connect() {
        return new Promise((resolve, reject) => {
            try {
                this.socket.connect(this.config.port, this.config.ip, () => {
                    this.isConnected = true;
                    this.emit('connected');
                    resolve();
                });
                this.socket.once('error', (err) => {
                    reject(err);
                });
            }
            catch (error) {
                if (error instanceof Error) {
                    reject(error);
                }
                else {
                    reject(new Error(String(error)));
                }
            }
        });
    }
    async send(data) {
        return new Promise((resolve, reject) => {
            if (!this.isConnected) {
                return reject(new Error('Socket not connected'));
            }
            this.socket.send(data, (err) => {
                if (err)
                    return reject(err);
                resolve();
            });
        });
    }
    close() {
        try {
            this.socket.close();
        }
        catch {
        }
    }
}
